File: src/args.ts
export type ParsedArgs = {
  flags: Set<string>;
  unknown: string[];
  help: boolean;
};

export function parseArgs(argv: string[], allowedFlags: Set<string>): ParsedArgs {
  const flags = new Set<string>();
  const unknown: string[] = [];
  let help = false;

  for (const arg of argv) {
    if (typeof arg !== "string" || arg.trim() === "") {
      continue;
    }

    if (arg === "--help" || arg === "-h") {
      help = true;
      continue;
    }

    if (arg.startsWith("--")) {
      const [flag] = arg.split("=", 1);
      if (allowedFlags.has(flag)) {
        flags.add(flag);
      } else {
        unknown.push(flag);
      }
      continue;
    }

    if (arg.startsWith("-") && arg.length > 1) {
      const shorts = arg.slice(1).split("");
      for (const short of shorts) {
        const flag = `-${short}`;
        if (allowedFlags.has(flag)) {
          flags.add(flag);
        } else {
          unknown.push(flag);
        }
      }
      continue;
    }

    unknown.push(arg);
  }

  return { flags, unknown, help };
}

File: src/args.go
package main

import (
	"strings"
)

// ParsedArgs holds the result of parsing command-line arguments
type ParsedArgs struct {
	Flags   map[string]bool
	Unknown []string
	Help    bool
}

// parseArgs parses command-line arguments
func parseArgs(argv []string, allowedFlags map[string]bool) ParsedArgs {
	flags := make(map[string]bool)
	unknown := []string{}
	help := false

	for _, arg := range argv {
		if arg == "" || strings.TrimSpace(arg) == "" {
			continue
		}

		if arg == "--help" || arg == "-h" {
			help = true
			continue
		}

		if strings.HasPrefix(arg, "--") {
			// Long flag
			parts := strings.SplitN(arg, "=", 2)
			flag := parts[0]
			if allowedFlags[flag] {
				flags[flag] = true
			} else {
				unknown = append(unknown, flag)
			}
			continue
		}

		if strings.HasPrefix(arg, "-") && len(arg) > 1 {
			// Short flags
			shorts := strings.Split(arg[1:], "")
			for _, short := range shorts {
				flag := "-" + short
				if allowedFlags[flag] {
					flags[flag] = true
				} else {
					unknown = append(unknown, flag)
				}
			}
			continue
		}

		unknown = append(unknown, arg)
	}

	return ParsedArgs{
		Flags:   flags,
		Unknown: unknown,
		Help:    help,
	}
}
